[Refactoring] Support refactoring calls to async if a variable or function is used as completion handler

ahoppen opened the pull request:
Previously, we only supported  refactoring a function to call the async alternative if a closure was used for the callback parameter. With this change, we also support calling a arbitrary function (or variable with function type) that is passed to the completion handler argument.

The implementation basically re-uses the code we already have to create the legacy function’s body (which calls the newly created async version and then forwards the arguments to the legacy completion handler).

To describe the completion handler that the result is being forwarded to, I’m also using `AsyncHandlerDesc`, but since the completion handler may be a variable, it doesn’t necessarily have an `Index` within a function decl that declares it. Because of this, I split the `AsyncHandlerDesc` up into a context-free `AsyncHandlerDesc` (without an `Index`) and `AsyncHandlerParamDesc` (which includes the `Index`). It turns out that `AsyncHandlerDesc` is sufficient in most places.


Resolves rdar://77460524 [SR-14573]

-------

This PR consists of three commits, which might be easiest to review on their own. After each commit, the source should be in a buildable state that passes all tests

### 1. Merge `LegacyAlternativeBodyCreator` into `AsyncConverter`

This will later allow us to reuse parts of `LegacyAlternativeBodyCreator` from `AsyncConverter` when refactoring calls to an async alternative if they pass a variable as the completion handler.

### 2. Replace usage of constant strings with `tok::` wherever possible

The code moved from `LegacyAlternativeBodyCreator` was using constant strings a lot. Use `tok::` instead to match the style of `AsyncConverter`.

### 3. Support refactoring calls to async if a variable or function is used as completion handler

The main change described above.

ahoppen committed:
[Refactoring] Merge `LegacyAlternativeBodyCreator` into `AsyncConverter`

This will later allow us to reuse parts of `LegacyAlternativeBodyCreator` from `AsyncConverter` when refactoring calls to an async alternative if they pass a variable as the completion handler.

ahoppen committed:
[Refactoring] Replace usage of constant strings with `tok::` wherever possible

The code moved from `LegacyAlternativeBodyCreator` was using constant strings a lot. Use `tok::` instead to match the style of `AsyncConverter`.

ahoppen committed:
[Refactoring] Support refactoring calls to async if a variable or function is used as completion handler

Previously, we only supported  refactoring a function to call the async alternative if a closure was used for the callback parameter. With this change, we also support calling a arbitrary function (or variable with function type) that is passed to the completion handler argument.

The implementation basically re-uses the code we already have to create the legacy function’s body (which calls the newly created async version and then forwards the arguments to the legacy completion handler).

To describe the completion handler that the result is being forwarded to, I’m also using `AsyncHandlerDesc`, but since the completion handler may be a variable, it doesn’t necessarily have an `Index` within a function decl that declares it. Because of this, I split the `AsyncHandlerDesc` up into a context-free `AsyncHandlerDesc` (without an `Index`) and `AsyncHandlerParamDesc` (which includes the `Index`). It turns out that `AsyncHandlerDesc` is sufficient in most places.

Resolves rdar://77460524

ahoppen commented:
@swift-ci Please test

bnbarham commented on test/refactoring/ConvertAsync/basic.swift:
```diff
@@ -656,3 +656,209 @@ func testCalls() {
 }
 // CONVERT-FUNC: {{^}}}
+
+func testSimpleWithVariableCompletionHandler(completionHandler: (String) -> Void) {
+  // RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=SIMPLE-WITH-VARIABLE-COMPLETION-HANDLER %s
+  simple(completion: completionHandler)
+  // SIMPLE-WITH-VARIABLE-COMPLETION-HANDLER: let result = await simple()
+  // SIMPLE-WITH-VARIABLE-COMPLETION-HANDLER: completionHandler(result)
+}
```
Also, if these *were* treated as completion handler functions, what happens when converting/adding async? eg.
```
func simple(_ completionHandler: () -> Void) {}

// Convert/add on `delay`
func delay(_ function: () -> Void) {
  simple(function)
}
```

ahoppen replied:
> IMO I'd name completionHandler something else for these tests, since otherwise they look like completion handler functions, but we're using convert call instead of convert function. maybe runAfter or something?

`testSimpleWithVariableCompletionHandler` is essentially just a wrapper around `simple`. The user might want to do some logging here, convert arguments etc, so I think `completionHandler` is the correct name. I 
deliberately didn’t use `completion` to make sure we are not picking up the variable name from the `simple` function declaration.

> The function as-is looks like it should just be converted to

I don’t think I follow. Is this the same point as the one below?

> Also, if these were treated as completion handler functions, what happens when converting/adding async? eg.

Good point. With this change, we rewrite
```swift
// RUN: %refactor -convert-to-async -dump-rewritten -source-filename %s -pos=%(line+1):1
func testSimpleWithVariableCompletionHandler(completionHandler: (String) -> Void) {
  simple(completion: completionHandler)
}
```

to 

```swift
func testSimpleWithVariableCompletionHandler() async -> String {
  let result = await simple()
  completionHandler(result)
}
```

which is definitely not correct.

ahoppen commented on test/refactoring/ConvertAsync/basic.swift:
```diff
@@ -656,3 +656,209 @@ func testCalls() {
   // VOID-AND-ERROR-CALL4-NEXT: {{^}}print("void and error completion \(<#v#>)"){{$}}
 }
 // CONVERT-FUNC: {{^}}}
+
```
I can split it in a follow-up PR once your and my changes are merged.

ahoppen committed:
[Refactoring] Support creation of async legacy bodies even if a parameter is not optional

If the parameter is not an `Optional`, add a placeholder instead that the user can fill with a sensible default value.

ahoppen committed:
[Refactoring] Promote call to refactored completion handlers to `return`

When a function’s completion handler is being passed to another function and the function that the completion handler is being declared in is converted to async, replace the call to the completion handler by a `return` statement.

For example:
```swift
func foo(completion: (String) -> Void) {
  bar(completion)
}
```

becomes
```swift
func foo() async -> String {
  return await bar()
}
```

Previously, we were calling the completion handler, which no longer exists in the newly created `async` function.

ahoppen committed:
[Refactoring] Check that code compiles after refatorings in `variable_as_callback.swift`

ahoppen commented:
OK, so I’ve moved all my new test cases to `variable_as_callback.swift`, so that `basic.swift` doesn’t get larger still. All other changes are in subsequent commits (hope this makes it easier to review)
- An improvement to the creation of legacy bodies: If a parameter to the completion handler is non-optional, don’t bail out entirely, but put in a placeholder that the user can fill with a sensible default value. This is important because we’re using the same logic to also convert calls with variables as completion handlers and don’t want to stop the refactoring entirely if one of the completion handler’s arguments is not `Optional`
- The case discussed above where a call to a removed completion handler should be promoted to a `return` statement. For example
```swift
func foo(completion: (String) -> Void) {
  bar(completion)
}
```
now becomes 
```swift
func foo() async -> String {
  return await bar()
}
```
- Some cleanup to the test (rename `CHECK` strings and check that the refactored code compiles)

bnbarham commented on test/refactoring/ConvertAsync/variable_as_callback.swift:
```diff
@@ -24,186 +24,320 @@ func genericResult<T>(completion: (T?, Error?) -> Void) where T: Numeric { }
 }
+// NO-PARAMS-VARIABLE-COMPLETION-HANDLER-FUNC: func testNoParamsVariableCompletionHandler() async {
+// NO-PARAMS-VARIABLE-COMPLETION-HANDLER-FUNC-NOT: return
+// NO-PARAMS-VARIABLE-COMPLETION-HANDLER-FUNC-NEXT:  await noParams
```
Should this be `noParams()`?

ahoppen replied:
Yes, it should.

bnbarham reviewed (approved):
Other than that one comment, looks good to me! You're probably going to beat me to merge, I realised there was another bug with the shadowing change - it wasn't renaming hoisted decls when the call was eg. within an if :(. So I'll probably be rebasing :)

ahoppen committed:
[Refactoring] Remove the `VARIABLE-COMPLETION-HANDLER` suffix from checks in `variable_as_callback.swift`

These were a legacy from when the tests lived in `basic.swift`. They are no longer needed.

ahoppen commented:
@swift-ci Please test
